### PR TITLE
Remove all LINQ usage from repo

### PR DESCRIPTION
**Pull request by kasperk81**
made usings explicit and worked out all linq dependencies.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [1.2.2] - 2024-05-23
```
Comment (baywet) [inline CHANGELOG.md:10]: number should be 1.2.4

File: src/FormSerializationWriter.cs [hunk 2]
```diff
@@ -162,18 +173,21 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
             value.Serialize(this);
         }
-        foreach(var additionalValueToMerge in additionalValuesToMerge.Where(static x => x != null))
+        foreach (var additionalValueToMerge in additionalValuesToMerge)
         {
-            OnBeforeObjectSerialization?.Invoke(additionalValueToMerge!);
-            OnStartObjectSerialization?.Invoke(additionalValueToMerge!, this);
-            additionalValueToMerge!.Serialize(this);
-            OnAfterObjectSerialization?.Invoke(additionalValueToMerge!);
+            if (additionalValueToMerge != null)
```
Comment (baywet) [inline src/FormSerializationWriter.cs:184]: nit, revert the condition, use continue, and remove the added curlies to avoid the arrow head shape of the code

File: src/Microsoft.Kiota.Serialization.Form.csproj [hunk 3]
```diff
@@ -8,7 +8,6 @@
     <Authors>Microsoft</Authors>
     <TargetFrameworks>netstandard2.0;netstandard2.1;net5.0;net6.0;net8.0</TargetFrameworks>
```
Comment (baywet) [inline src/Microsoft.Kiota.Serialization.Form.csproj:9]: the version needs to be bumped

File: src/FormParseNode.cs [hunk 4]
```diff
@@ -184,17 +204,26 @@ public IEnumerable<T> GetCollectionOfPrimitiveValues<T>()
...
     {
-        foreach (var v in DecodedValue.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries))
-            yield return GetEnumValueInternal<T>(v);
+        List<T?> enumValues = new List<T?>();
+        string[] values = DecodedValue.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
+        foreach (string v in values)
+        {
+            enumValues.Add(GetEnumValueInternal<T>(v));
+        }
+        return enumValues;
```
Comment (baywet) [inline src/FormParseNode.cs:223]: I'm not sure I understand why this change is necessary?

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for the contributions!
A couple of remarks.

File: src/FormParseNode.cs (hunk 4 above)
Comment (filipnavara) [reply src/FormParseNode.cs:223]: +1 ... It's not necessary. `yield return` generates a perfectly fine enumerator instance.

File: src/FormParseNode.cs [hunk 5]
```diff
@@ -208,7 +237,8 @@ public IEnumerable<T> GetCollectionOfPrimitiveValues<T>()
     {
         if(string.IsNullOrEmpty(rawValue))
             return null;
-        if(typeof(T).GetCustomAttributes<FlagsAttribute>().Any())
+
+        if (typeof(T).GetCustomAttributes<FlagsAttribute>().GetEnumerator().MoveNext())
```
Comment (filipnavara) [inline src/FormParseNode.cs:241]: I thought about doing this. The catch is that the enumerator may be `IDisposable`. I didn't actually verify if it is or not but the `Any()` pattern handles it.

File: src/FormSerializationWriterFactory.cs [hunk 6]
```diff
@@ -1,4 +1,5 @@
 using Microsoft.Kiota.Abstractions.Serialization;
+using System;
```
Comment (filipnavara) [inline src/FormSerializationWriterFactory.cs:2]: Unnecessary change?

File: src/FormParseNode.cs (hunk 4 above)
Comment (filipnavara) [reply src/FormParseNode.cs:223]: There's no LINQ in the method already. It's a generated iterator which is not problematic per se.

File: src/FormParseNode.cs (hunk 5 above)
Comment (baywet) [reply src/FormParseNode.cs]: good catch!

File: src/FormSerializationWriter.cs [hunk 7]
```diff
@@ -206,24 +226,25 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
     {
-        if(values == null) return;
+        if (values is null) return;
 
-        StringBuilder? valueNames = null;
-        foreach(var x in values)
+        var enumerator = values.GetEnumerator();
+
+        if (enumerator is IDisposable de) de.Dispose();
+        if (!enumerator.MoveNext()) return;
```
Comment (filipnavara) [inline src/FormSerializationWriter.cs:234]: You should not dispose the enumerator before using it.

File: src/FormParseNode.cs [hunk 8]
```diff
@@ -208,7 +231,14 @@ public IEnumerable<T> GetCollectionOfPrimitiveValues<T>()
...
             return null;
-        if(typeof(T).GetCustomAttributes<FlagsAttribute>().Any())
+
+        bool hasFlagsAttribute = false;
+        using (var enumerator = typeof(T).GetCustomAttributes<FlagsAttribute>().GetEnumerator())
+        {
+            hasFlagsAttribute = enumerator.MoveNext();
+        }
+
+        if (hasFlagsAttribute)
```
Comment (MichalStrehovsky) [inline src/FormParseNode.cs:241]: This avoids creating a garbage instance of the attribute and enumerator as well (same comment to the other occurrence of this pattern in this PR):

```suggestion
        if (typeof(T).IsDefined(typeof(FlagsAttribute)))
```

File: src/FormSerializationWriter.cs [hunk 9]
```diff
@@ -162,18 +173,27 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
         depth++;
-        if(value == null && !additionalValuesToMerge.Any(static x => x != null)) return;
+
+        if (value == null)
+        {
+            var enumerator = additionalValuesToMerge.GetEnumerator();
+            bool hasAdditionalValues = enumerator.MoveNext();
+            if (enumerator is IDisposable de) de.Dispose();
+            if (!hasAdditionalValues) return;
+        }
```
Comment (MichalStrehovsky) [inline src/FormSerializationWriter.cs:183]: Is this equivalent to the previous version? This looke like `Any()` without a predicate.

It might help readability to just define your own tiny LINQ with the obvious implementation and use that. E.g. https://github.com/dotnet/corert/blob/master/src/Common/src/Internal/LowLevelLinq/LowLevelEnumerable.cs

The problem with the real LINQ is that it's not implemented the way the "obvious" implementation would be. It's _a lot_ more complicated.

File: src/FormSerializationWriter.cs [hunk 10]
```diff
@@ -233,13 +254,14 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
     {
         if(value.HasValue)
         {
-            if(typeof(T).GetCustomAttributes<FlagsAttribute>().Any())
+            using var enumerator = typeof(T).GetCustomAttributes<FlagsAttribute>().GetEnumerator();
+            if (enumerator.MoveNext())
```
Comment (MichalStrehovsky) [inline src/FormSerializationWriter.cs:258]: Use IsDefined(typeof(FlagsAttribute)) too?

File: src/FormSerializationWriter.cs [hunk 11]
```diff
@@ -162,18 +173,27 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
         depth++;
-        if(value == null && !additionalValuesToMerge.Any(static x => x != null)) return;
+
+        if (value == null)
+        {
+            using (var enumerator = additionalValuesToMerge.GetEnumerator())
+            {
+                if (!enumerator.MoveNext()) return;
+            }
+        }
```
Comment (MichalStrehovsky) [inline src/FormSerializationWriter.cs:183]: I think this is still different from the original. There should be a foreach loop looking for x != null.

File: src/FormSerializationWriter.cs [hunk 12]
```diff
@@ -162,18 +173,27 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
         if(depth > 0) throw new InvalidOperationException("Form serialization does not support nested objects.");
         depth++;
-        if(value == null && !additionalValuesToMerge.Any(static x => x != null)) return;
+        if (value == null)
+        {
+            foreach (var x in additionalValuesToMerge)
+            {
+                if (x != null) break;
+            }
+        }
```
Comment (baywet) [inline src/FormSerializationWriter.cs:182]:
```suggestion
        if (value == null && !Array.Exists(additionalValuesToMerge, static x => x is not null)) return;
```

File: src/FormSerializationWriter.cs [hunk 13]
```diff
@@ -206,24 +226,25 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
+        using (var enumerator = values.GetEnumerator())
         {
-            if(x.HasValue && Enum.GetName(typeof(T), x.Value) is string valueName)
+            if (!enumerator.MoveNext()) return;
+        }
+
+        List<string> nonNullValues = new List<string>();
+        foreach (var value in values)
+        {
+            if (value.HasValue)
```
Comment (baywet) [inline src/FormSerializationWriter.cs:239]: I'm not sure why we changed that condition? This will break enums serialization where the symbol name is different from the wirename

**Review by baywet — CHANGES_REQUESTED** [review]
